clang/lib/CodeGen/CGDebugInfo.cpp

```diff
@@ -2021,28 +2021,10 @@ llvm::DISubroutineType *CGDebugInfo::getOrCreateInstanceMethodType(
   // ThisPtr may be null if the member function has an explicit 'this'
   // parameter.
   if (!ThisPtr.isNull()) {
-    const CXXRecordDecl *RD = ThisPtr->getPointeeCXXRecordDecl();
-    if (isa<ClassTemplateSpecializationDecl>(RD)) {
```

Comment:
Is removing this entire block NFC? I'm surprised that this doesn't affect anything. If so, we should probably make the NFC change separately from the HLSL specific one so that it's clear what (isn't) changing.

Reply:
Well, this is a consequence of the `this` fix, the same done in DXC. After applying the fix the code becomes:

```c++
if (isa<ClassTemplateSpecializationDecl>(RD)) {
    // HLSL Change Begin - This is a reference.
    llvm::DIType *ThisPtrType = getOrCreateType(ThisPtr, Unit);
    // HLSL Change End - This is a reference.
    TypeCache[ThisPtr.getAsOpaquePtr()].reset(ThisPtrType);
    // TODO: This and the artificial type below are misleading, the
    // types aren't artificial the argument is, but the current
    // metadata doesn't represent that.
    ThisPtrType = DBuilder.createObjectPointerType(ThisPtrType);
    Elts.push_back(ThisPtrType);
  } else {
    llvm::DIType *ThisPtrType = getOrCreateType(ThisPtr, Unit);
    TypeCache[ThisPtr.getAsOpaquePtr()].reset(ThisPtrType);
    ThisPtrType = DBuilder.createObjectPointerType(ThisPtrType);
    Elts.push_back(ThisPtrType);
  }
```

Removing the comments, if and else blocks become the same code, so I just removed the branching

Reply:
This code has been around in clang for _a long time_, and it was implemented back when LLVM's testing standards were not quite as high as they are today (https://github.com/llvm/llvm-project/commit/fa59ac363ef4747fdb142d993326c43d832629bb).

The test case that was introduced with the original code here is [here](https://github.com/llvm/llvm-project/blob/main/clang/test/CodeGenCXX/debug-info-template-limit.cpp), and it generates identical IR with and without this PR.

It would be nice to have someone who understands C++ debug info well review this change as well, but it looks to me like the special handling here is vestigial and no longer needed.

@dwblaikie, @adrian-prantl, or @Michael137, can one of you take a look at this please?

Reply:
huh, interesting archaeology... Wish we knew which test in particular it was.
This change might've been before I changed one of the old/core (before vtables, template specializations, and ctor homing) -flimit-debug-info thing which was more ad-hoc, to be powered by class "required to be complete" events. 

So, yeah, I could totally believe whatever this code did may've been subsumed by other work/generalizations/etc since then.

but, yeah, if we're removing it, best to do it separately in an NFC change & if someone's feeling like doing due diligence, perhaps building clang with/without this change (on linux, or otherwise with -flimit-debug-info enabled) and checking the binary output is identical?

Reply:
Created a NFC with this change, please take a look here: https://github.com/llvm/llvm-project/pull/119445